libclamav/udf.c

```diff
@@ -0,0 +1,838 @@
+    IMPLEMENTATION_USE_VOLUME_DESCRIPTOR        = 4,
+    LOGICAL_VOLUME_DESCRIPTOR                   = 6,
+    PARTITION_DESCRIPTOR                        = 5,
+    UNALLOCATED_SPACE_DESCRIPTOR                = 7,
+    TERMINATING_DESCRIPTOR                      = 8,
+    LOGICAL_VOLUME_INTEGRITY_DESCRIPTOR         = 9,
+    ANCHOR_VOLUME_DESCRIPTOR_DESCRIPTOR_POINTER = 2,
+    FILE_SET_DESCRIPTOR                         = 256
+} VOLUME_DESCRIPTOR_TAG;
+
+static PrimaryVolumeDescriptor *getPrimaryVolumeDescriptor(cli_ctx *ctx, size_t *idxp, size_t *lastOffsetp)
+{
+    uint8_t *buffer               = NULL;
+    PrimaryVolumeDescriptor *test = NULL;
+    PrimaryVolumeDescriptor *ret  = NULL;
+    size_t idx                    = *idxp;
+    size_t lastOffset             = *lastOffsetp;
+
+    if (skipEmptyDescriptors(ctx, idxp, lastOffsetp)) {
+        goto done;
+    }
+
+    idx        = *idxp;
+    lastOffset = *lastOffsetp;
+
+    buffer = (uint8_t *)fmap_need_off(ctx->fmap, idx, VOLUME_DESCRIPTOR_SIZE);
+    if (NULL == buffer) {
+        goto done;
+    }
+    lastOffset = idx;
+
+    test = (PrimaryVolumeDescriptor *)buffer;
+    if (PRIMARY_VOLUME_DESCRIPTOR != test->tag.tagId) {
+        goto done;
+    }
+
+    idx += VOLUME_DESCRIPTOR_SIZE;
+    ret = test;
+
+done:
+    *idxp        = idx;
+    *lastOffsetp = lastOffset;
+
+    return ret;
+}
+
+static ImplementationUseVolumeDescriptor *getImplementationUseVolumeDescriptor(cli_ctx *ctx, size_t *idxp, size_t *lastOffsetp)
+{
+    uint8_t *buffer                         = NULL;
+    ImplementationUseVolumeDescriptor *test = NULL;
+    ImplementationUseVolumeDescriptor *ret  = NULL;
+    size_t idx                              = *idxp;
+    size_t lastOffset                       = *lastOffsetp;
+
+    if (skipEmptyDescriptors(ctx, idxp, lastOffsetp)) {
+        goto done;
+    }
+
+    idx        = *idxp;
+    lastOffset = *lastOffsetp;
+
+    buffer = (uint8_t *)fmap_need_off(ctx->fmap, idx, VOLUME_DESCRIPTOR_SIZE);
+    if (NULL == buffer) {
+        goto done;
+    }
+    lastOffset = idx;
+
+    test = (ImplementationUseVolumeDescriptor *)buffer;
+    if (IMPLEMENTATION_USE_VOLUME_DESCRIPTOR != test->tag.tagId) {
+        goto done;
+    }
+
+    ret = test;
+    idx += VOLUME_DESCRIPTOR_SIZE;
+
+done:
+    *idxp        = idx;
+    *lastOffsetp = lastOffset;
+
+    return ret;
+}
+
+static LogicalVolumeDescriptor *getLogicalVolumeDescriptor(cli_ctx *ctx, size_t *idxp, size_t *lastOffsetp)
+{
+    uint8_t *buffer               = NULL;
+    LogicalVolumeDescriptor *ret  = NULL;
+    LogicalVolumeDescriptor *test = NULL;
+    size_t idx                    = *idxp;
+    size_t lastOffset             = *lastOffsetp;
+
+    if (skipEmptyDescriptors(ctx, idxp, lastOffsetp)) {
+        goto done;
+    }
+
+    idx        = *idxp;
+    lastOffset = *lastOffsetp;
+
+    buffer = (uint8_t *)fmap_need_off(ctx->fmap, idx, VOLUME_DESCRIPTOR_SIZE);
```

Comment:
I'm a little concerned if this needs to be unneeded. Or maybe use `fmap_need_off_once()` instead